zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -67,18 +95,39 @@ impl Chain {
 
     /// Fork a chain at the block with the given hash, if it is part of this
     /// chain.
-    pub fn fork(&self, fork_tip: block::Hash) -> Option<Self> {
+    ///
+    /// `finalized_tip_history_tree`: the history tree for the finalized tip
+    /// from which the tree of the fork will be computed.
+    pub fn fork(
+        &self,
+        fork_tip: block::Hash,
+        finalized_tip_history_tree: &HistoryTree,
+    ) -> Result<Option<Self>, HistoryTreeError> {
         if !self.height_by_hash.contains_key(&fork_tip) {
-            return None;
+            return Ok(None);
         }
 
-        let mut forked = self.clone();
+        let mut forked = self.with_history_tree(finalized_tip_history_tree.clone());
 
         while forked.non_finalized_tip_hash() != fork_tip {
             forked.pop_tip();
         }
 
-        Some(forked)
+        // Rebuild the history tree starting from the finalized tip tree.
+        // TODO: change to a more efficient approach by removing nodes
+        // from the tree of the original chain (in `pop_tip()`).
```

Comment:
Task tracking: Can you open a ticket for this follow-up optimisation?

Reply:
This looks like #2378 - can you add the ticket number to this comment?

Reply:
I've added it in the line below in 030240f and forgot to comment here. (Also it seems GitHub is being weird and is not showing it in the conversation tab)